File: scala/learning-ci-scala/src/main/scala/learningci/chapter02/Critics.scala
package learningci.chapter02

case class Person(val name: String) {
}

object Person {
  val lisa = Person("Lisa Rose")
  val gene = Person("Gene Seymour")
  val michael = Person("Michael Phillips")
  val claudia = Person("Claudia Puig")
  val mick = Person("Mick LaSalle")
  val jack = Person("Jack Matthews")
  val toby = Person("Toby")
}

case class Movie(val title: String) {
}

object Movie {
  val ladyInTheWater = Movie("Lady in the Water")
  val snakesOnAPlane = Movie("Snakes on a Plane")
  val justMyLuck = Movie("Just My Luck")
  val supermanReturns = Movie("Superman Returns")
  val youMeAndDupree = Movie("You, Me and Dupree")
  val theNightListener = Movie("The Night Listener")
}

case class Critic(val person: Person, val movie: Movie, val rating: Double) {
}

object Critics {

  import Person._
  import Movie._

  val all: List[Critic] = List(

    Critic(lisa, ladyInTheWater, 2.5D)
    , Critic(lisa, snakesOnAPlane, 3.5D)
    , Critic(lisa, justMyLuck, 3.0D)
    , Critic(lisa, supermanReturns, 3.5D)
    , Critic(lisa, youMeAndDupree, 2.5D)
    , Critic(lisa, theNightListener, 3.0D)

    , Critic(gene, ladyInTheWater, 3.0D)
    , Critic(gene, snakesOnAPlane, 3.5D)
    , Critic(gene, justMyLuck, 1.5D)
    , Critic(gene, supermanReturns, 5.0D)
    , Critic(gene, youMeAndDupree, 3.5D)
    , Critic(gene, theNightListener, 3.0D)

    , Critic(michael, ladyInTheWater, 2.5D)
    , Critic(michael, snakesOnAPlane, 3.0D)
    , Critic(michael, supermanReturns, 3.5D)
    , Critic(michael, theNightListener, 3.5D)

    , Critic(claudia, snakesOnAPlane, 3.5D)
    , Critic(claudia, justMyLuck, 3.0D)
    , Critic(claudia, supermanReturns, 4.0D)
    , Critic(claudia, youMeAndDupree, 2.5D)
    , Critic(claudia, theNightListener, 4.5D)

    , Critic(mick, ladyInTheWater, 3.0D)
    , Critic(mick, snakesOnAPlane, 4.0D)
    , Critic(mick, justMyLuck, 2.0D)
    , Critic(mick, supermanReturns, 3.0D)
    , Critic(mick, youMeAndDupree, 2.0D)
    , Critic(mick, theNightListener, 3.0D)

    , Critic(jack, ladyInTheWater, 3.0D)
    , Critic(jack, snakesOnAPlane, 4.0D)
    , Critic(jack, supermanReturns, 5.0D)
    , Critic(jack, youMeAndDupree, 3.5D)
    , Critic(jack, theNightListener, 3.0D)

    , Critic(toby, snakesOnAPlane, 4.5D)
    , Critic(toby, supermanReturns, 4.0D)
    , Critic(toby, youMeAndDupree, 1.0D)

  )

}

File: ruby/lib/chapter02/critics.rb
#!/usr/bin/env ruby

module Chapter02

  module NameHolder
    def initialize(name)
      @name = name
    end 
    attr_reader :name 
    def to_s() @name end
  end

  class Person 
    include NameHolder

    LisaRose        = Person.new('Lisa Rose')
    GeneSeymour     = Person.new('Gene Seymour')
    MichaelPhillips = Person.new('Michael Phillips')
    ClaudiaPuig     = Person.new('Claudia Puig')
    MickLaSalle     = Person.new('Mick LaSalle')
    JackMatthews    = Person.new('Jack Matthews')
    Toby            = Person.new('Toby')
  end

  class Movie 
    include NameHolder

    LadyInTheWater   = Movie.new('Lady in the Water')
    SnakesOnAPlane   = Movie.new('Snakes on a Plane')
    JustMyLuck       = Movie.new('Just My Luck')
    SupermanReturns  = Movie.new('Superman Returns')
    YouMeAndDupree   = Movie.new('You, Me and Dupree')
    TheNightListener = Movie.new('The Night Listener')
  end

  class Critic
    def initialize(person,movie,rating)
      @person = person
      @movie = movie
      @rating = rating
    end
    attr_reader :person, :movie, :rating
    def to_s() 
      "(" + @person.to_s + "," + @movie.to_s + "," + @rating.to_s + ")"
    end
  end

  module Critics
    All = Array.new([
      Critic.new(Person::LisaRose, Movie::LadyInTheWater,   2.5),
      Critic.new(Person::LisaRose, Movie::SnakesOnAPlane,   3.5),
      Critic.new(Person::LisaRose, Movie::JustMyLuck,       3.0),
      Critic.new(Person::LisaRose, Movie::SupermanReturns,  3.5),
      Critic.new(Person::LisaRose, Movie::YouMeAndDupree,   2.5),
      Critic.new(Person::LisaRose, Movie::TheNightListener, 3.0),
      Critic.new(Person::GeneSeymour, Movie::LadyInTheWater,   3.0),
      Critic.new(Person::GeneSeymour, Movie::SnakesOnAPlane,   3.5),
      Critic.new(Person::GeneSeymour, Movie::JustMyLuck,       1.5),
      Critic.new(Person::GeneSeymour, Movie::SupermanReturns,  5.0),
      Critic.new(Person::GeneSeymour, Movie::YouMeAndDupree,   3.5),
      Critic.new(Person::GeneSeymour, Movie::TheNightListener, 3.0),
      Critic.new(Person::MichaelPhillips, Movie::LadyInTheWater,   2.5),
      Critic.new(Person::MichaelPhillips, Movie::SnakesOnAPlane,   3.0),
      Critic.new(Person::MichaelPhillips, Movie::SupermanReturns,  3.5),
      Critic.new(Person::MichaelPhillips, Movie::TheNightListener, 3.5),
      Critic.new(Person::ClaudiaPuig, Movie::SnakesOnAPlane,   3.5),
      Critic.new(Person::ClaudiaPuig, Movie::JustMyLuck,       3.0),
      Critic.new(Person::ClaudiaPuig, Movie::SupermanReturns,  4.0),
      Critic.new(Person::ClaudiaPuig, Movie::YouMeAndDupree,   2.5),
      Critic.new(Person::ClaudiaPuig, Movie::TheNightListener, 4.5),
      Critic.new(Person::MickLaSalle, Movie::LadyInTheWater,   3.0),
      Critic.new(Person::MickLaSalle, Movie::SnakesOnAPlane,   4.0),
      Critic.new(Person::MickLaSalle, Movie::JustMyLuck,       2.0),
      Critic.new(Person::MickLaSalle, Movie::SupermanReturns,  3.0),
      Critic.new(Person::MickLaSalle, Movie::YouMeAndDupree,   2.0),
      Critic.new(Person::MickLaSalle, Movie::TheNightListener, 3.0),
      Critic.new(Person::JackMatthews, Movie::LadyInTheWater,   3.0),
      Critic.new(Person::JackMatthews, Movie::SnakesOnAPlane,   4.0),
      Critic.new(Person::JackMatthews, Movie::SupermanReturns,  5.0),
      Critic.new(Person::JackMatthews, Movie::YouMeAndDupree,   3.5),
      Critic.new(Person::JackMatthews, Movie::TheNightListener, 3.0),
      Critic.new(Person::Toby, Movie::SnakesOnAPlane,   4.5),
      Critic.new(Person::Toby, Movie::SupermanReturns,  4.0),
      Critic.new(Person::Toby, Movie::YouMeAndDupree,   1.0),
    ])
  end

end

